ambry-router/src/main/java/com.github.ambry.router/PutOperation.java

```diff
@@ -810,13 +879,19 @@ private void prepareForSending() {
     /**
      * Do the actions required when the chunk has been completely built.
      * @param updateMetric whether chunk fill completion metrics should be updated.
+     * @param encryptContent {@code true} if chunk's content need to be encrypted. {@code false} otherwise
      */
-    void onFillComplete(boolean updateMetric) {
+    void onFillComplete(boolean updateMetric, boolean encryptContent) {
       buf.flip();
-      prepareForSending();
-      chunkReadyAtMs = time.milliseconds();
+      chunkBlobSize = buf.remaining();
       if (updateMetric) {
-        routerMetrics.chunkFillTimeMs.update(chunkReadyAtMs - chunkFreeAtMs);
+        routerMetrics.chunkFillTimeMs.update(time.milliseconds() - chunkFreeAtMs);
+      }
+      if (!passedInBlobProperties.isEncrypted()) {
+        prepareForSending();
+        chunkReadyAtMs = time.milliseconds();
+      } else {
+        encryptChunk(this, encryptContent);
```

Comment:
why is the `PutChunk` calling into a method in `PutOperation`? Usually, the operation class calls into the methods in the chunks - both for puts and gets. Can `encryptChunk()` be a method within `PutChunk`?

Reply:
Yep. Makes sense. Will fix it.